### DISCUSSION TITLE
Confused about the anonymous service account

### DISCUSSION THREAD

**Discussion by daenney** [kanidm/kanidm#2146] [Q&A] [answered]
I'm experimenting with Kani and so far it's been amazing. But there's some things about the anonymous service account that I can't quite figure out. I'd like to understand in which "systems" anonymous is active. Kani seems to strongly err on the side of safety and modern security primitives, so an anonymous user that can do somethings by default was very surprising.

I'm familiar with LDAP's anonymous bind for login, but that can be avoided through a service account instead. Based on looking at the code, it also seems that anonymous is never allowed for Oauth2? So that leaves whatever the PAM module is doing?

---

I also think I ran into a few bugs figuring things out as I go along. I'm not entirely sure if this is me misunderstanding the docs and CLI interactions, but I figured I should include them.

Going through the documentation, I also came across this example: https://github.com/kanidm/kanidm/blob/6d67464a60264107d5a3a734487ef734e459a7aa/book/src/accounts_and_groups.md#L93-L99

However, that doesn't work for me. I always get:
```
~ ❯❯❯ kanidm login --name anonymous
Login Success for anonymous@idm.example.com
~ ❯❯❯ kanidm person get test --name anonymous
2023-09-25T12:52:16.240657Z ERROR kanidm_cli::common: Privileges have expired for anonymous@idm.example.com - you need to re-authenticate again.
Would you like to re-authenticate? yes
2023-09-25T12:52:17.768295Z ERROR kanidm_cli::session: Error during reauthentication begin phase: Http(500, Some(InvalidState), "46d5ba68-0fc3-4335-9455-4d7b619908e8")
```

On the server side I see:

```
46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     request [ 439µs | 22.47% / 100.00% ]
46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     ┝━ reauth [ 337µs | 34.05% / 76.83% ]
46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     │  ┝━ ｉ [info]: Begin reauth event | event_tag_id: 10
46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     │  ┝━ validate_and_parse_token_to_ident [ 188µs | 42.78% ]
46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     │  ┝━ ｉ [info]: Initiating Re-Authentication Session | event_tag_id: 10 | username: anonymous | issue: Token | uuid: 00000000-0000-0000-0000-ffffffffffff
46d5ba68-0fc3-4335-9455-4d7b619908e8 ERROR    │  ┝━ 🚨 [error]: Ident session is not present in entry. Perhaps replication is delayed?
46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     │  ┕━ ｉ [info]: Sending reauth result | event_tag_id: 10 | res: Err(InvalidState)
46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     ┝━ auth_session_state_management [ 3.08µs | 0.70% ] inter: Err(InvalidState)
46d5ba68-0fc3-4335-9455-4d7b619908e8 ERROR    ┕━ 🚨 [error]: response failed | classification: Status code: 500 Internal Server Error | latency: 0 ms
00000000-0000-0000-0000-000000000000 ERROR    🚨 [error]: Failed to complete connection: hyper::Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
```

Once I set up a pair of posix credentials, then I can:
```
~ ❯❯❯ kanidm person posix show test --name anonymous
---
spn: test@idm.example.com
name: test
displayname: Test
uuid: 286348a4-f25a-476d-86cd-f3a1d246c984
shell: <none>
group: [ spn: test@idm.example.com, gidnumber: 3527854468 name: test, uuid: 286348a4-f25a-476d-86cd-f3a1d246c984 ]
```

But `person get test --name anonymous` keeps failing.

I figured that since I don't need this currently, I'd at least go in and lock the `anonymous` service account. But that also leads to some confusion for me:

```
~ ❯❯❯ kanidm service-account validity show anonymous --name idm_admin
user: anonymous
valid after: any time
expire: never
```

This looks fine, until I try an invalid user:
```
~ ❯❯❯ kanidm service-account validity show doesnotexist --name idm_admin
user: doesnotexist
valid after: any time
expire: never
```

I would expect a non-existing user to return empty or some error here.

I then proceeded to try and lock the anonymous one:

```
~ ❯❯❯ kanidm login --name idm_admin
Enter password: 
Login Success for idm_admin@idm.example.com
~ ❯❯❯ kanidm service-account validity expire-at anonymous epoch --name idm_admin
2023-09-25T12:58:30.784158Z ERROR kanidm_cli::serviceaccount: Error -> ParseFromDescription(InvalidComponent("year"))
```

The same error is returned when I try to use "now", or any of the other ones like "any". But using a datetime does work:
```
~ ❯❯❯ kanidm service-account validity expire-at anonymous 2023-01-01T00:00:00+00:00 --name idm_admin
Success
```

And now indeed, I can no longer login as anonymous, which also means the `anonymous` posix command no longer works:
```
~ ❯❯❯ kanidm person posix show test --name anonymous
2023-09-25T13:01:12.524916Z ERROR kanidm_cli::person: Error -> Http(401, Some(SessionExpired), "822a41e6-d66a-43c9-a695-4e43899de96d")
```
```
~ ❯❯❯ kanidm login --name anonymous
2023-09-25T13:00:28.451548Z ERROR kanidm_cli::session: Error during authentication init phase: AuthenticationFailed
```

System info:
```
~ ❯❯❯ kanidm version
kanidm 1.1.0-beta.13
2023-09-25T13:04:04.698228Z DEBUG kanidm: Using 16 worker threads

~ ❯❯❯ kanidmd version
kanidmd 1.1.0-beta.13
```

**Comment by Firstyear** [comment, marked as answer]
> I'm experimenting with Kani and so far it's been amazing.

Thank you! 

> But there's some things about the anonymous service account that I can't quite figure out. I'd like to understand in which "systems" anonymous is active. Kani seems to strongly err on the side of safety and modern security primitives, so an anonymous user that can do somethings by default was very surprising.

We are starting to phase it out. Many opensource LDAP deployments rely on and use anonymous access for some public data access that is not infringing on personal information. Anonymous in Kani is "just enough" to make unix-authentication work for these purposes. 

> 
> I'm familiar with LDAP's anonymous bind for login, but that can be avoided through a service account instead. Based on looking at the code, it also seems that anonymous is never allowed for Oauth2? So that leaves whatever the PAM module is doing?

There are two issues related to this. See #2027 and #41 - both are being worked on at the moment. 

And as you have identified, yes, the anonymous user is really only for pam/nss purposes. In time we may add support for machine accounts which will allow us to cease the use of anonymous in these spaces. 

Part of the reason we don't "mind" anonymous is that if you have a few hundred servers, you need a service-account for each one. If you "share" a service account between all of them, you have a shared password and may as well just have an anonymous system for all intents and purposes.

So to do this properly we need a way to have machine accounts, and that's been delayed for now due to a lack of need. However, for some future integrations we likely will need to support machine accounts so this will happen in the future. 

> 
> I also think I ran into a few bugs figuring things out as I go along. I'm not entirely sure if this is me misunderstanding the docs and CLI interactions, but I figured I should include them.
> 
> Going through the documentation, I also came across this example:
> 
> https://github.com/kanidm/kanidm/blob/6d67464a60264107d5a3a734487ef734e459a7aa/book/src/accounts_and_groups.md#L93-L99
> 
> However, that doesn't work for me. I always get:
> 
> ```
> ~ ❯❯❯ kanidm login --name anonymous
> Login Success for anonymous@idm.example.com
> ~ ❯❯❯ kanidm person get test --name anonymous
> 2023-09-25T12:52:16.240657Z ERROR kanidm_cli::common: Privileges have expired for anonymous@idm.example.com - you need to re-authenticate again.
> Would you like to re-authenticate? yes
> 2023-09-25T12:52:17.768295Z ERROR kanidm_cli::session: Error during reauthentication begin phase: Http(500, Some(InvalidState), "46d5ba68-0fc3-4335-9455-4d7b619908e8")
> ```

Interesting - I think it's a documentation error that we list "person get" here, it should be "service-account get". 

The request for privileges also is possibly a bug, I'd need to check why we did that. Would you mind opening the bug report? 

The reason anonymous fails to access privileges is that we have session tokens in a "read-only" or "read-write" state, where the read-write state requires re-authentication and has a time limit. Anonymous as a security feature is never allowed to access the read-write state for example.




> 
> On the server side I see:
> 
> ```
> 46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     request [ 439µs | 22.47% / 100.00% ]
> 46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     ┝━ reauth [ 337µs | 34.05% / 76.83% ]
> 46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     │  ┝━ ｉ [info]: Begin reauth event | event_tag_id: 10
> 46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     │  ┝━ validate_and_parse_token_to_ident [ 188µs | 42.78% ]
> 46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     │  ┝━ ｉ [info]: Initiating Re-Authentication Session | event_tag_id: 10 | username: anonymous | issue: Token | uuid: 00000000-0000-0000-0000-ffffffffffff
> 46d5ba68-0fc3-4335-9455-4d7b619908e8 ERROR    │  ┝━ 🚨 [error]: Ident session is not present in entry. Perhaps replication is delayed?
> 46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     │  ┕━ ｉ [info]: Sending reauth result | event_tag_id: 10 | res: Err(InvalidState)
> 46d5ba68-0fc3-4335-9455-4d7b619908e8 INFO     ┝━ auth_session_state_management [ 3.08µs | 0.70% ] inter: Err(InvalidState)
> 46d5ba68-0fc3-4335-9455-4d7b619908e8 ERROR    ┕━ 🚨 [error]: response failed | classification: Status code: 500 Internal Server Error | latency: 0 ms
> 00000000-0000-0000-0000-000000000000 ERROR    🚨 [error]: Failed to complete connection: hyper::Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
> ```
> 
> Once I set up a pair of posix credentials, then I can:
> 
> ```
> ~ ❯❯❯ kanidm person posix show test --name anonymous
> ---
> spn: test@idm.example.com
> name: test
> displayname: Test
> uuid: 286348a4-f25a-476d-86cd-f3a1d246c984
> shell: <none>
> group: [ spn: test@idm.example.com, gidnumber: 3527854468 name: test, uuid: 286348a4-f25a-476d-86cd-f3a1d246c984 ]
> ```

This one is also a bug, I think we might not be limiting the posix get here to persons & posix, but may just be showing all posix capable accounts ... 

> 
> But `person get test --name anonymous` keeps failing.
> 
> I figured that since I don't need this currently, I'd at least go in and lock the `anonymous` service account. But that also leads to some confusion for me:
> 
> ```
> ~ ❯❯❯ kanidm service-account validity show anonymous --name idm_admin
> user: anonymous
> valid after: any time
> expire: never
> ```
> 
> This looks fine, until I try an invalid user:
> 
> ```
> ~ ❯❯❯ kanidm service-account validity show doesnotexist --name idm_admin
> user: doesnotexist
> valid after: any time
> expire: never
> ```
> 
> I would expect a non-existing user to return empty or some error here.

That is also very curious! 

> 
> I then proceeded to try and lock the anonymous one:
> 
> ```
> ~ ❯❯❯ kanidm login --name idm_admin
> Enter password: 
> Login Success for idm_admin@idm.example.com
> ~ ❯❯❯ kanidm service-account validity expire-at anonymous epoch --name idm_admin
> 2023-09-25T12:58:30.784158Z ERROR kanidm_cli::serviceaccount: Error -> ParseFromDescription(InvalidComponent("year"))
> ```
> 
> The same error is returned when I try to use "now", or any of the other ones like "any". But using a datetime does work:

Also a bug! Now/any/never etc should work. 

> 
> ```
> ~ ❯❯❯ kanidm service-account validity expire-at anonymous 2023-01-01T00:00:00+00:00 --name idm_admin
> Success
> ```
> 
> And now indeed, I can no longer login as anonymous, which also means the `anonymous` posix command no longer works:
> 
> ```
> ~ ❯❯❯ kanidm person posix show test --name anonymous
> 2023-09-25T13:01:12.524916Z ERROR kanidm_cli::person: Error -> Http(401, Some(SessionExpired), "822a41e6-d66a-43c9-a695-4e43899de96d")
> ```

Hey, at the least when the account lock *did* succeed it did lock! So we got the important part correct! 

It looks mostly like some small inconsistencies with the CLI and the docs, so please open an issue so we can fix these items :) 

And thanks for raising the discussion, we are always happy to help,

**Reply by daenney** [reply]
Brilliant, thanks for the response! I'll keep the anonymous one disabled, since I'm not planning to use the Unix/LDAP auth for now. I think the one difference where the SA might be nice, is that if you did expose Kani to the outside world (as you might need to do for the Oauth2 bits), then things like anon bind couldn't be used to attempt some username enumeration. But that's a fairly minor thing in the end, because knowing a username exists shouldn't grant you anything useful. I guess that's also what #611 would help counter.

I'll raise issues for the CLI interactions!

**Reply by Firstyear** [reply]
Yes, that is absolutely a consideration - it's why we made the anonymous account "just like any other account" where you can disable to prevent this. And of course, as you note knowledge of a username shouldn't give you anything anyway, so that's been factored in too :) 

The ip access controls are probably going to have to happen soon too, we've been slowly getting the infrastructure in place for them, and I think it would be possible to do these soon. We simply have too many ideas and not enough time :) 

And thanks for the many issues about the cli interactions - we'll get onto those as that kind of polish is really good for the project.
